File: 0645. Set Mismatch/0645.java
// time complexity: O(n)
// space complexity: O(n)
class Solution {
    public int[] findErrorNums(int[] nums) {
        // first index will be zero always
        boolean[] seen = new boolean[nums.length+1];

        // find duplicate
        int duplicate = 0;
        for(int num : nums){
            if(seen[num]) duplicate = num;
            else seen[num] = true;
        }

        // find missing
        int missing = 0;
        for(int i=1; i<=nums.length; i++){
            if(!seen[i]){
                missing = i;
                break;
            }
        }

        return new int[]{duplicate, missing};
    }
}

File: 0645. Set Mismatch/0645.py
# time complexity: O(n)
# space complexity: O(n)
class Solution(object):
    def findErrorNums(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        # expected sum
        expected_sum = n*(n + 1) // 2 
        # actual sum
        actual_sum = sum(nums)  # O(n) time, O(1) space
        
        # find the duplicate number
        duplicate = None
        seen = set()
        for num in nums:
            if num in seen:
                duplicate = num  # duplicate found
            else:
                seen.add(num)
        
        # calculate the missing number
        missing = expected_sum - (actual_sum - duplicate)
        
        return [duplicate, missing]